### DISCUSSION TITLE
Pickable not working when supplying attributes directly

### DISCUSSION THREAD

**Discussion by osctro** [visgl/deck.gl#7503] [Q&A] [answered]
I made a plain javascript PoC for tracking a set of vehicles on a map using the ScatterPlotLayer. Server polling is made in a worker thread that uses postMessage to send the indices array (plain js array of integers) and the positions array (typed float32) to the map.

In order to update the positions I am following the https://deck.gl/docs/developer-guide/performance#supply-attributes-directly and everything works as expected except when I enable pickable that makes the points disappear.

The main code is:
```
const RECORD_LEN = 2
const POLLING_WORKER = new Worker("polling-worker.js");
const DEVICE_DATA = {length: 0, attributes: {}}
let lastUpdateTime = Date.now();
let glContext;

const TILE_LAYER = new deck.TileLayer({
    data: "https://c.tile.openstreetmap.org/{z}/{x}/{y}.png",
    minZoom: 0,
    maxZoom: 19,
    renderSubLayers: props => {
        const { bbox: {west, south, east, north} } = props.tile;
        return new deck.BitmapLayer(props, {data: null, image: props.data, bounds: [west, south, east, north]});
    }
});

const DECK = new deck.DeckGL({
    initialViewState: {longitude: 23.319941, latitude: 42.698334, zoom: 7, maxZoom: 16},
    controller: true,
    getTooltip: ({object}) => object && "Hi from map!!!",
    onWebGLInitialized
});

function onWebGLInitialized(gl) {
    glContext = gl;
    render()
}

function render() {
    const deviceLayer = new deck.ScatterplotLayer({
        id: "devices",
        data: DEVICE_DATA,
        stroked: false,
        filled: true,
        radiusMinPixels: 3,
        getRadius: 8,
        getFillColor: [255, 0, 0],
        updateTriggers: {getPosition: lastUpdateTime},
        pickable: true,
        autoHighlight: true,
        onHover: (e) => console.log("Hovered:", e),
        getTooltip: ({object}) => object && "Hi from layer!!!"
    });
    DECK.setProps({layers: [TILE_LAYER, deviceLayer]})
}

POLLING_WORKER.onmessage = (e) => {
    if (!glContext) return;
    
   const positions = e.data[1];
    DEVICE_DATA.length = positions.length / RECORD_LEN; 
    const indices = e.data[0]
    console.log(indices)
    const buffer = new luma.Buffer(glContext, positions)
    DEVICE_DATA.attributes= {
        indices,
        getPosition: {buffer, type: 0x1406, size: 2}
    };
    
   lastUpdateTime = Date.now();
    render();
}
POLLING_WORKER.postMessage('init');
```

If I add an additional ScatterPlotLayer using JSON data and accessors it works so it seems a problem when supplying attrs directly. Could you give me some advice on how to solve it?.

BTW, what's the role of the indices array? Now I am always updating the full positions array and setting indices array from 0 to array len -1. Could it be possible to send only the positions that have changed from one server request to next and use the indices array to let deck.gl know which entries in the complete buffer are being passed.

Thank you.

**Comment by Pessimistress** [comment, marked as answer]
> what's the role of the indices array? 

ScatterplotLayer does not use any attribute called indices. You can remove it.

A layer's data prop is only processed when its value shallow changes. You should not mutate the same data object on updates.
